servicetalk-dns-discovery-netty/src/main/java/io/servicetalk/dns/discovery/netty/DnsServiceDiscovererObserver.java

```diff
@@ -46,6 +60,18 @@ interface DnsDiscoveryObserver {
          * @return {@link DnsResolutionObserver} that provides visibility into results of the current DNS resolution
          */
         DnsResolutionObserver onNewResolution(String name);
+
+        /**
+         * Notifies that the current DNS discovery got canceled (did not complete successfully).
```

Comment:
Consider describing to users the expected behavior of terminal callbacks. See `HttpLifecycleObserver` as an example:
https://github.com/apple/servicetalk/blob/b4543f4a42111758903d9fdbcd795b1be4b71a61/servicetalk-http-api/src/main/java/io/servicetalk/http/api/HttpLifecycleObserver.java#L104-L110

https://github.com/apple/servicetalk/blob/b4543f4a42111758903d9fdbcd795b1be4b71a61/servicetalk-http-api/src/main/java/io/servicetalk/http/api/HttpLifecycleObserver.java#L146-L153